### ISSUE TITLE
Unsafe type Inference with spread

### ISSUE THREAD

**Issue by changchanghwang** [microsoft/TypeScript#60878]
### 🔎 Search Terms

`unsafe type with spread`, `spread`, `optional type`, `type inference`

### 🕗 Version & Regression Information

I think all versions

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/MYGwhgzhAECCBCBhaBvAUNaBLAJgLmggBcAnLAOwHMBuDaMA4sq2u4Ae3KYFdgj2SACjAlKEAinqNSFStAC+ASlR1MRABZYIAOlzQAvNABEARiO1MazTrAH6omxYVo6AdywbhDifQD805jklFUtoDi4iegAjYDtyAFNXOCRBdFDMbUyNLQAaVUsGezFtMDzQpSdMMBjdHDtsnVxK6BJ4om4ScmjgJ3k0PrRw4m64xOTEVIYAcnV4kBB2aFcBEBwppUHOCHYQeO0FykEjVohuECI8Ixzu7XdPFCUSxSA

### 💻 Code

```ts
class ABC {
  id: string;
  a: string;

  constructor(args: { a: string }) {
    this.id = "1";
    this.a = args.a;
  }

  with(args: { a?: string }) {
    const abc = new ABC({
      ...this,
      a: args.a,
    });
    abc.id = this.id;
    return abc;
  }
}
```

### 🙁 Actual behavior

when I use `with` method without arguements like below
```ts
const abc = new ABC({ a: 'hello world' });
const newAbc = abc.with({});
```

newAbc.a is undefined.
constructor want `a is required` but with give constructor undefined.
I think it's not expected working.

### 🙂 Expected behavior

It should be caught by compiler. because It's not developer expected working!

**Comment by jcalz** [comment]
Note: not a TS team member.

Duplicate #60623, #50559, #10727. `this` is effectively generic, generic spread is modeled by intersections, which don't adequately capture how properties can be overwritten. 

The title and search terms do not adequately summarize or reflect the issue you're facing. If this were not a duplicate, I'd suggest you edit to address it.

**Comment by changchanghwang** [comment]
> Note: not a TS team member.
> 
> Duplicate [#60623](https://github.com/microsoft/TypeScript/issues/60623), [#50559](https://github.com/microsoft/TypeScript/issues/50559), [#10727](https://github.com/microsoft/TypeScript/issues/10727). `this` is effectively generic, generic spread is modeled by intersections, which don't adequately capture how properties can be overwritten.
> 
> The title and search terms do not adequately summarize or reflect the issue you're facing. If this were not a duplicate, I'd suggest you edit to address it.

sorry, I'm not good at english🥲
is it related with generic? 
<img width="330" alt="Image" src="https://github.com/user-attachments/assets/3f3ad412-c5a5-4d0c-8ce9-d359f5810135" />

In this manner, if it becomes `this & { a: string | undefined; }`, then `a` can ultimately be undefined, so I think it should be inferred as optional, regardless of what `this` is.

**Comment by MartinJohns** [comment]
> I think it should be inferred as optional

But it's not optional, it's **always** provided. Optional has different semantics (defined or not defined), which matters with `exactOptionalPropertyTypes` enabled.

**Comment by changchanghwang** [comment]
When TypeScript uses the spread operator, excess property checks can be unintentionally disabled, resulting in unsafe type inference. 
It seems like we need an option to fix this issue. 
Could we handle it similarly to how object literals are type-checked?
